### PR TITLE
added list initialisation

### PR DESCRIPTION
**Pull request by gf712**
list initialisation to replace starting empty vector and the replace elements

### CODE REVIEW

File: examples/meta/generator/targets/octave.json [hunk 1]
```diff
@@ -12,7 +12,6 @@
             "InitialSeperatorWhenArgs>0": false
         },
         "string": "$name = $arguments",
-        "CharVector": "$name = zeros(1, $arguments, 'char')",
```
Comment (gf712) [inline examples/meta/generator/targets/octave.json]: this doesn't work in octave (4.4.1)

File: examples/meta/generator/targets/ruby.json [hunk 2]
```diff
@@ -33,6 +33,18 @@
...
+        "ByteVector": "$name = NArray['byte', [$arguments])",
+        "WordVector": "$name = NArray['byte', [$arguments])",
+        "IntVector": "$name = NArray['sint', [$arguments])",
+        "LongIntVector": "$name = NArray['int', [$arguments])",
+        "ULongIntVector": "$name = NArray['int', [$arguments])",
+        "ShortRealVector": "$name = NArray['sfloat', [$arguments]))",
+        "RealVector": "$name = NArray['float', [$arguments])",
+        "LongRealVector": "$name = NArray['float', [$arguments])",
+        "ComplexVector": "NArray['scomplex', [$arguments])"
+    },
```
Comment (gf712) [inline examples/meta/generator/targets/ruby.json]: I tested this out with the most recent numo/NArray version

Comment (gf712) [reply examples/meta/generator/targets/ruby.json]: ah actually, what type are the zero constructors? I can't find it anywhere

File: examples/meta/generator/translate.py [hunk 3]
```diff
@@ -458,6 +461,42 @@ def translateStatement(self, statement):
...
+        typeKey = list(typeDict.keys())[0]
+
+        self.variableTypes[nameString] = typeDict
+
+        if list(initialisation.keys())[0] == "ArgumentList":
+            template = Template(self.targetDict["Init"]["Construct"])
+            # custom SGType construction
+            if typeKey == "ShogunSGType"\
+                and init[0][typeKey] in self.targetDict["ListInit"]:
+                template = Template(self.targetDict["ListInit"][init[0][typeKey]])
```
Comment (gf712) [inline examples/meta/generator/translate.py]: This should probably be moved out of the condition. This substitution only works if there is an implementation of the list initialiser of the target language

File: examples/meta/generator/targets/csharp.json [hunk 4]
```diff
@@ -32,6 +32,19 @@
...
     },
+    "ListInit": {
+        "BoolVector": "var $name = new bool[] {$arguments}",
+        "CharVector": "var $name = new char[] {$arguments}",
+        "ByteVector": "var $name = new byte[] {$arguments}",
+        "WordVector": "var $name = new ushort[] {$arguments}",
+        "IntVector": "var $name = new int[] {$arguments}",
+        "LongIntVector": "var $name = new long[] {$arguments}",
+        "ULongIntVector": "var $name = new ulongint[] {$arguments}",
+        "ShortRealVector": "var $name = new float[] {$arguments}",
```
Comment (gf712) [inline examples/meta/generator/targets/csharp.json]: need to fix c#, floats need to have a 'f' suffix, i.e. 1.0, should become 1.0f ...

File: examples/meta/generator/targets/java.json [hunk 5]
```diff
@@ -38,6 +38,18 @@
         "RealMatrix": "DoubleMatrix $name = new DoubleMatrix($arguments)",
         "LongRealMatrix": "DoubleMatrix $name = new DoubleMatrix($arguments)"
     },
+    "ListInit": {
+        "BoolVector": "BoolVector $name = new BoolVector({$arguments})",
```
Comment (gf712) [inline examples/meta/generator/targets/java.json]: not sure how to do this one

Comment (karlnapf) [reply examples/meta/generator/targets/java.json]: Don’t know either, last resort would be to translate it to explicit assignment although that would suck

**Comment by gf712** [comment]
> Like it! I guess now it’s done work to make it work with all the langs...
> 
> It would be cool if you could also try to make it work with R and ruby. While those meta examples are currently disabled in the build, they in principle work (r doesn’t have static calls, ruby doesn’t have overloaded methods/functions) since both support list inits. I think it would be sufficient to look at generates output and locally try it that works

OK, I can have a look into R. With ruby I am not sure how it would work, as the old NArray doesn't seem to have a list initialiser? The new NArray library does seem to have it though. But I am not sure which one shogun supports?

File: examples/meta/generator/targets/java.json [hunk 6]
```diff
@@ -34,10 +34,22 @@
         "IntMatrix": "DoubleMatrix $name = new DoubleMatrix($arguments)",
         "LongIntMatrix": "DoubleMatrix $name = new DoubleMatrix($arguments)",
         "ULongIntMatrix": "DoubleMatrix $name = new DoubleMatrix($arguments)",
-        "ShortRealMatrix": "DoubleMatrix $name = new DoubleMatrix($arguments)",
+        "ShortRealMatrix": "FloatMatrix $name = new FloatMatrix($arguments)",
```
Comment (gf712) [inline examples/meta/generator/targets/java.json]: I am assuming this should be `FloatMatrix`?

**Comment by karlnapf** [comment]
> I figured out how to make it work, but there is no support for bools in ruby's old NArray library, so it might be worth upgrading to numo-array

Shogun still uses the old one and we have an issue for upgrading [here](#4178). It is some work as the typemaps will have to be re-written. That is cumbersome as ruby is currently disabled due to #4177.
So don't loose your sleep over this I'd say. Or

File: examples/meta/generator/targets/java.json (hunk 6 above)
Comment (karlnapf) [reply examples/meta/generator/targets/java.json]: not sure what the typemap exactly does...but I think our current Java bindings treat everything as double matrix. Might be worth cleaning that up at some point soon

**Comment by gf712** [comment]
> Shogun still uses the old one and we have an issue for upgrading [here](#4178). It is some work as the typemaps will have to be re-written. That is cumbersome as ruby is currently disabled due to #4177.
> So don't loose your sleep over this I'd say. Or

I am thinking that maybe I skip boolean for now? There is still some lack of support for it in java, ruby and r (at least) so can address that separately?

File: examples/meta/generator/targets/ruby.json (hunk 2 above)
Comment (karlnapf) [reply examples/meta/generator/targets/ruby.json]: what do you mean?

File: examples/meta/src/meta_api/list_init.sg [hunk 7]
```diff
@@ -1,5 +1,3 @@
-BoolVector bool_vector([True, False])
```
Comment (karlnapf) [inline examples/meta/src/meta_api/list_init.sg]: ok!

**Comment by karlnapf** [comment]
Soooo merge? :)

**Comment by gf712** [comment]
Yup, seems fine to me. When there is a need I can add matrix initialisation

File: examples/meta/generator/targets/ruby.json (hunk 2 above)
Comment (gf712) [reply examples/meta/generator/targets/ruby.json]: sorry, this was when I was confused as to how to construct with a list of args in ruby. was simpler than I thought! :D

**Pull request merged**